gnat2goto/driver/arrays.adb

```diff
@@ -117,49 +114,116 @@ package body Arrays is
+              Make_Dereference_Expr (Object          => Lhs_Ptr,
+                                     Source_Location => Source_Loc,
+                                     I_Type          => Element_Type);
+         begin
+            Append_Declare_And_Init (Symbol     => Literal_Temp,
+                                     Value      => Malloc_Call_Expr,
+                                     Block      => Result_Block,
+                                     Source_Loc => Source_Loc);
+
+            --  Handle an "others" splat expression if present:
+            if Present (Component_Associations (N)) then
+               declare
+                  Maybe_Others_Node : constant Node_Id :=
+                    Last (Component_Associations (N));
+                  Maybe_Others_Choices : constant List_Id :=
+                    Choices (Maybe_Others_Node);
+               begin
+                  pragma Assert (List_Length (Maybe_Others_Choices) = 1);
+
+                  --  this association does not end with others -> bail
+                  if Nkind (First (Maybe_Others_Choices)) /= N_Others_Choice
+                  then
+                     return;
+                  end if;
+
+                  Others_Expression :=
+                    Do_Expression (Expression (Maybe_Others_Node));
+               end;
+            else
+               return;
             end if;
 
-            declare
-               Others_Node : constant Node_Id :=
-                 First (Component_Associations (N));
-               Others_Choices : constant List_Id := Choices (Others_Node);
-               Expr : constant Irep :=
-                 Do_Expression (Expression (Others_Node));
-            begin
-               if List_Length (Others_Choices) /= 1 then
-                  return Report_Unhandled_Node_Irep (N,
-                                                 "Do_Aggregate_Literal_Array",
-                                                 "More than one other choice");
-               end if;
-               if Nkind (First (Others_Choices)) /= N_Others_Choice then
-                  return Report_Unhandled_Node_Irep (N,
-                                                 "Do_Aggregate_Literal_Array",
-                                                 "Wrong kind of other choice");
-               end if;
-               Array_Expr :=
-                 Make_Op_Array_Of (I_Type => Bare_Array_Type,
-                                   Op0 => Expr,
-                                   Source_Location => Sloc (N));
-            end;
-            With_Mode := True;
-         else
-            Array_Expr := Make_Array_Expr (I_Type => Bare_Array_Type,
-                                           Source_Location => Sloc (N));
-            With_Mode := False;
-         end if;
+            --  iterate over elements and assing others-value to them
+            Append_Op (Loop_Body,
+                       Make_Code_Assign (Rhs             => Others_Expression,
```

Comment:
Nvm, I get why the initialisation is like this now, but the `Others_Expression` should still be declared closer to here IMHO, and the logic should probably be flipped (early exit if there are no component associations, early exit with "not handled yet" diagnostic if there are component associations other than `others`)

Reply:
It was much smaller in the beginning (before the whole loop-init) but now I agree: the declaration should take place once we know we are handling the right case.